apollo-federation/src/supergraph/mod.rs

```diff
@@ -96,16 +98,19 @@ pub(crate) fn extract_subgraphs_from_supergraph(
     let (mut subgraphs, federation_spec_definitions, graph_enum_value_name_to_subgraph_name) =
         collect_empty_subgraphs(supergraph_schema, join_spec_definition)?;
 
-    let mut filtered_types = Vec::new();
-    for type_definition_position in supergraph_schema.get_types() {
-        if !join_spec_definition
-            .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?
-            && !link_spec_definition
-                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?
-        {
-            filtered_types.push(type_definition_position);
-        }
-    }
+    let filtered_types: Vec<_> = supergraph_schema
+        .get_types()
+        .fallible_filter(|type_definition_position| {
+            join_spec_definition
+                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())
+                .map(Not::not)
+        })
+        .and_then_filter(|type_definition_position| {
+            link_spec_definition
+                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())
+                .map(Not::not)
+        })
```

Comment:
`.map(not)` feels very awkward to me, I think I prefer `?` +  `Ok`:
```suggestion
        .fallible_filter(|type_definition_position| {
            Ok(!join_spec_definition
                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?)
        })
        .and_then_filter(|type_definition_position| {
            Ok(!link_spec_definition
                .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?)
        })
```

Reply:
I agree with you, but there is a reason for this. The compiler can not infer what the error type is for the `Ok`s. This is true with the `fallible_filter` + `and_then_filter` and if the two filters combined into something like:
```rs
        .fallible_filter(|type_definition_position| {
            Ok(!join_spec_definition
                     .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?
                  ||  !link_spec_definition
                        .is_spec_type_name(supergraph_schema, type_definition_position.type_name())?
        )})
```